pkg/build/api/v1/types.go

```diff
@@ -163,6 +163,9 @@ type BuildSource struct {
 	// repository.
 	ContextDir string `json:"contextDir,omitempty" description:"specifies sub-directory where the source code for the application exists, allows for sources to be built from a directory other than the root of a repository"`
 
+	// DockerfileName is the name of the Dockerfile that will be used to build the Docker image.
+	DockerfileName string `json:"dockerfileName,omitempty" description:"specifies the name of the Dockerfile to use for building the Docker image"`
```

Comment:
>  Only things generic to all strategies should be in source.

we have the Dockerfile string field in source, which does not apply to s2i builds or, likely, custom builds.

The line between a filename being a "input" (like context dir or actual file content) and it being "part of a strategy's decision making process" is a pretty blurry imho.  Putting it in the source definition is saying "here are your inputs", the strategy can decide what do with those inputs, including throw an error because it doesn't make sense (as happens if you specify a Dockerfile string when running a source-strategy build)